<a href="https://colab.research.google.com/github/afviyanabila/ab-testing-marketing-campaign-analysis/blob/main/a_b_testing_marketing_campaign_analysis_afviya_nabila.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **A/B TESTING : FOOD MARKETING CAMPAIGN**


---

**by Afviya Nabila**

[Linkedin](https://www.linkedin.com/in/afviya-nabila/)
[Github](https://github.com/afviyanabila/)
[Slides](https://)

---



## **Project Overview**

A fast-food chain has a total of 137 stores with three market size types. It plans to add a new item to its menu. However, they are still undecided between three possible marketing campaigns for promoting the new item

In order to determine which promotion has the greatest effect on sales, the new item is introduced in the store in several randomly selected market types

Three different promotions implemented at different location, and the weekly sales of the new item are recorded for the first four weeks.The mean of the sales amount is 53.5 thousand dollars


## **Data Source**

*   Marketing Campaign Dataset

Reference : [Kaggle](https://www.kaggle.com/datasets/chebotinaa/fast-food-marketing-campaign-ab-test)

## **Import Data**

In [8]:
import pandas as pd

data = "https://raw.githubusercontent.com/rajeevratan84/datascienceforbusiness/master/WA_Fn-UseC_-Marketing-Campaign-Eff-UseC_-FastF.csv"
market_campaign = pd.read_csv(data)

## **Data Preparation & Data Cleaning**

In [9]:
market_campaign.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 548 entries, 0 to 547
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   MarketID          548 non-null    int64  
 1   MarketSize        548 non-null    object 
 2   LocationID        548 non-null    int64  
 3   AgeOfStore        548 non-null    int64  
 4   Promotion         548 non-null    int64  
 5   week              548 non-null    int64  
 6   SalesInThousands  548 non-null    float64
dtypes: float64(1), int64(5), object(1)
memory usage: 30.1+ KB


In [10]:
market_campaign.shape

(548, 7)

In [11]:
print('\n data missing values :', market_campaign.isna().sum().sum())
market_campaign.drop_duplicates(inplace=True)
print('\n data duplicated values :', market_campaign.duplicated().sum())


 data missing values : 0

 data duplicated values : 0


The statistic of each aspects of the dataset also provided

In [15]:
market_campaign.describe().T

count        mean         std    min      25%    50%  \
MarketID          548.0    5.715328    2.877001   1.00    3.000    6.0   
LocationID        548.0  479.656934  287.973679   1.00  216.000  504.0   
AgeOfStore        548.0    8.503650    6.638345   1.00    4.000    7.0   
week              548.0    2.500000    1.119055   1.00    1.750    2.5   
SalesInThousands  548.0   53.466204   16.755216  17.34   42.545   50.2   

                       75%     max  
MarketID            8.0000   10.00  
LocationID        708.0000  920.00  
AgeOfStore         12.0000   28.00  
week                3.2500    4.00  
SalesInThousands   60.4775   99.65

Marketing campaign dataset consists of 548 entries and 7 columns without any missing and duplicate values

---

*   MarketId: market type id identified as number (1-20)
*   MarketSize: market size types (small, medium, large)
*   LocationID: store location id identified as number (1-920)
*   AgeOfStore : the running age of the stores (1-28 years)
*   Promotion: the promotion that were tested identified as number (1, 2, 3)
*   Week: the week when the promotion were run (1-4)
*   SalesinThousands: sales amount from the promotion implemented in the store within a specified week
---


For simplicity, the name of contents of the promotion being changed
*   1 = Promotion_1
*   2 = Promotion_2
*   3 = Promotion_3

In [12]:
market_campaign['Promotion'] = [f'Promotion_{value}' for value in market_campaign['Promotion']]

Here's the head of marketing campaign dataset:

In [13]:
market_campaign.head()

MarketID MarketSize  LocationID  AgeOfStore    Promotion  week  \
0         1     Medium           1           4  Promotion_3     1   
1         1     Medium           1           4  Promotion_3     2   
2         1     Medium           1           4  Promotion_3     3   
3         1     Medium           1           4  Promotion_3     4   
4         1     Medium           2           5  Promotion_2     1   

   SalesInThousands  
0             33.73  
1             35.67  
2             29.03  
3             39.25  
4             27.81

## **Exploratory Data Analysis**


In [16]:
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
import scipy.stats as stats

#### **Sales Distribution by Different Promotions**

In [ ]:
fig_sales_pie = px.pie(market_campaign, values=market_campaign['SalesInThousands'], color=market_campaign['Promotion'], names=market_campaign['Promotion'], color_discrete_sequence=['#d00000', '#ffba08', '#3f88c5'], hole=0.25)
fig_sales_pie.update_traces(textposition='inside', textfont=dict(color='white',size=15), textinfo='label+percent+value',pull=[0.1,0,0,0,0.1],rotation = 90)
fig_sales_pie.update_layout(title='<b>Sales Distribution by Different Promotions</b>', title_font=dict(size=20))

fig_sales_pie.show()

The highest sales distribution comes from Promotion_3 covered 35.5% of total sales with 10,408.52 thousand dollars followed by Promotion_1 and Promotion_2 being the least with less than 10,000 thousand dollars in four weeks promotion

In [28]:
sales_stat = market_campaign.groupby('Promotion').describe()['SalesInThousands']

sales_stat

count       mean        std    min      25%     50%      75%  \
Promotion                                                                   
Promotion_1  172.0  58.099012  16.553782  30.81  46.3525  55.385  63.6175   
Promotion_2  188.0  47.329415  15.108955  17.34  38.1700  45.385  51.7450   
Promotion_3  188.0  55.364468  16.766231  22.18  44.1975  51.165  61.7325   

               max  
Promotion           
Promotion_1  99.65  
Promotion_2  88.64  
Promotion_3  96.48

The majority of stores earned below 51.7 to 63.6 thousand dollars from the implemented promotions. Conversely, the average earnings covered a range of about 47.3 to 58.1 thousand dollars

Notably from statistics, Promotion_1 held the maximum and the highest mean of earnings, yet Promotion_3 accounted for the greatest total earnings due to its wider implementation across markets

#### **Market Size Distribution in Different Promotions**

In [35]:
m_size = market_campaign['MarketSize'].value_counts().reset_index()
m_size.columns = ['MarketSize', 'Counts']

m_size

MarketSize  Counts
0     Medium     320
1      Large     168
2      Small      60

In [19]:
fig_msize_bar = px.bar(m_size, x='MarketSize', y='Counts', color='MarketSize', color_discrete_sequence=['#ffba08', '#3f88c5', '#d00000'])
fig_msize_bar.update_traces(textangle=0, textposition="inside")
fig_msize_bar.update_layout(title='<b>Market Size Distribution</b>', title_font=dict(size=20), yaxis_title='Counts')

fig_msize_bar.show()

In [ ]:
market_campaign['MarketSize'].value_counts().sum()

548

The fast food chain has a total of 548 markets with more than half being the medium market size

The relationship between promotions and market size then showed by pivoting market size data

In [20]:
market_campaign.groupby(['Promotion', 'MarketSize']).count()['MarketID'].unstack('MarketSize')

MarketSize   Large  Medium  Small
Promotion                        
Promotion_1     56      96     20
Promotion_2     64     108     16
Promotion_3     48     116     24

In [21]:
marketsize_promo = market_campaign.groupby(['Promotion','MarketSize']).agg(count=('MarketSize','count')).sort_values(by=['count']).reset_index()
fig_msize_sunburst = px.sunburst(marketsize_promo, path=['Promotion','MarketSize'], values='count', color_discrete_sequence=['#ffba08', '#d00000', '#3f88c5'])
fig_msize_sunburst.update_traces(textinfo='label+value', rotation = 120)
fig_msize_sunburst.update_layout(title='<b>Market Size Distribution in Different Promotions</b>', title_font=dict(size=20))

fig_msize_sunburst.show()

The distribution of market size among each promotion appears to be quite similar, particularly between Promotion_3 and Promotion_2. This suggests that the promotions are being conducted in a relatively balanced manner across the different market sizes

#### **Age of Store in Different Promotions**

In [22]:
storeage = market_campaign.groupby(['AgeOfStore', 'Promotion']).count()['MarketID'].unstack('Promotion').sort_values(by='AgeOfStore', ascending=True)
storeage.head()

Promotion   Promotion_1  Promotion_2  Promotion_3
AgeOfStore                                       
1                  24.0         36.0         20.0
2                   8.0          8.0          4.0
3                  16.0         12.0          4.0
4                  16.0         12.0         16.0
5                   8.0         12.0         24.0

In [23]:
storeage_promo = market_campaign.groupby(['AgeOfStore', 'Promotion']).agg(count=('AgeOfStore','count')).reset_index()
fig_storeage_hist = px.histogram(storeage_promo, x='count', y='AgeOfStore', orientation='h',color='Promotion', color_discrete_sequence=['#3f88c5','#ffba08','#d00000'], barmode='group')
fig_storeage_hist.update_yaxes(tickmode='linear', dtick=5)
fig_storeage_hist.update_layout(title='<b>Age Of Store in Different Promotions</b>', title_font=dict(size=20), xaxis_title='Counts')

fig_storeage_hist.show()

In [37]:
market_campaign.groupby('Promotion').describe()['AgeOfStore']

count      mean       std  min  25%  50%   75%   max
Promotion                                                        
Promotion_1  172.0  8.279070  6.636160  1.0  3.0  6.0  12.0  27.0
Promotion_2  188.0  7.978723  6.597648  1.0  3.0  7.0  10.0  28.0
Promotion_3  188.0  9.234043  6.651646  1.0  5.0  8.0  12.0  24.0

The histogram is revealing that a significant portion of markets have an age of store ranging from 0-4 years for Promotion_1 and Promotion_2 and
5 to 9 years for Promotion_3.

The statistical analysis aligns with the histogram discloses that around 75% of the markets implementing promotions have a store age below 10-12 years, with an average age of approximately 8-9 years.

## **A/B Testing**

In order to compare the promotions, we need to implement A/B testing with two hypotheses concepts:
*   **Null Hypothesis (H0)**: The statement that there is no effect or no significant difference between variables
*   **Alternative Hypothesis (H1)**: The statement that contradicts the null hypothesis

and two type of values:
*   **T-value**: The value computed from the sample data that summarizes the evidence against the null hypothesis. *The larger value, the higher degree of difference of variables*
*   **P-value**: The value showed the probability of obtaining a test statistic as extreme as the one calculated from the sample data, assuming the null hypothesis is true. *The smaller value, the more statistically difference of variables*

The significant level (α) we used is 0.05%. Subsequently,
*   **P-value ≤ α**, you reject the null hypothesis and conclude that there is evidence to support the alternative hypothesis (significant difference)
*   **P-value > α**, you fail to reject the null hypothesis, meaning that there is not enough evidence to support the alternative hypothesis (no significant difference)

Histogram and Shapiro-Wilk test presented to check the distribution of sales by different promotions and whether each of them is normally distributed or not in an effort to choose the suited type of hypothesis

In [25]:
norm = px.histogram(market_campaign, x='SalesInThousands', color=market_campaign['Promotion'], color_discrete_sequence=['#d00000', '#ffba08', '#3f88c5'], nbins=20, facet_col='Promotion', facet_col_wrap=3)
norm.update_layout(title='<b>Sales Distribution by Different Promotions</b>', yaxis_title='Counts')

norm.show()

In [26]:
grouped_data = market_campaign.groupby('Promotion')['SalesInThousands']

for promotion, data in grouped_data:
    shapiro_statistic, p_value = stats.shapiro(data)
    print(f"Promotion: {promotion}")
    print("Shapiro-Wilk Test:", shapiro_statistic)
    print("P-value:", p_value)

    if p_value < 0.05:
        print("Data is not normally distributed")
    else:
        print("Data is normally distributed")
    print("\n")

Promotion: Promotion_1
Shapiro-Wilk Test: 0.9152998328208923
P-value: 1.9772498305314912e-08
Data is not normally distributed


Promotion: Promotion_2
Shapiro-Wilk Test: 0.9145088195800781
P-value: 5.456247009760773e-09
Data is not normally distributed


Promotion: Promotion_3
Shapiro-Wilk Test: 0.9207685589790344
P-value: 1.499518376135711e-08
Data is not normally distributed




Guided by the results above, histogram showed that all types of promotions are the combination of two distinct groups of data with double-peaked distribution which shows bimodal distribution which also confirmed by the Shapiro-Wilk Test whose pointed out that each promotion not normally distributed

For this reason, we can't applied the common t-test and utilize non-parametric test, such as Mann-Whitney U test

**Mann-Whitney U test** compares distributions of two independent groups without asumming normality (non-normally distributed, unequal variances)


### **Mann-Whitney U test**

#### **Promotion_1 vs Promotion_2**

In [ ]:
group1_data1 = market_campaign[market_campaign['Promotion'] == "Promotion_1"]["SalesInThousands"]
group1_data2 = market_campaign[market_campaign['Promotion'] == "Promotion_2"]["SalesInThousands"]

U_statistic, p_value = stats.mannwhitneyu(group1_data1, group1_data2, alternative='two-sided')

print("Mann-Whitney U statistic:", U_statistic)
print("P-value:", p_value)

if p_value < 0.01:
    print("Reject null hypothesis: Groups are different")
else:
    print("Fail to reject null hypothesis: No significant difference between groups")

Mann-Whitney U statistic: 22957.5
P-value: 5.845935246838518e-12
Reject null hypothesis: Groups are different


#### **Promotion_2 vs Promotion_3**

In [ ]:
group2_data2 = market_campaign[market_campaign['Promotion'] == "Promotion_2"]["SalesInThousands"]
group2_data3 = market_campaign[market_campaign['Promotion'] == "Promotion_3"]["SalesInThousands"]

U_statistic, p_value = stats.mannwhitneyu(group2_data2, group2_data3, alternative='two-sided')

print("Mann-Whitney U statistic:", U_statistic)
print("P-value:", p_value)

if p_value < 0.01:
    print("Reject null hypothesis: Groups are different")
else:
    print("Fail to reject null hypothesis: No significant difference between groups")

Mann-Whitney U statistic: 12093.0
P-value: 1.1970084441651803e-07
Reject null hypothesis: Groups are different


#### **Promotion_1 vs Promotion_3**

In [ ]:
group3_data1 = market_campaign[market_campaign['Promotion'] == "Promotion_1"]["SalesInThousands"]
group3_data3 = market_campaign[market_campaign['Promotion'] == "Promotion_3"]["SalesInThousands"]

U_statistic, p_value = stats.mannwhitneyu(group3_data1, group3_data3, alternative='two-sided')

print("Mann-Whitney U statistic:", U_statistic)
print("P-value:", p_value)

if p_value < 0.01:
    print("Reject null hypothesis: Groups are different")
else:
    print("Fail to reject null hypothesis: No significant difference between groups")

Mann-Whitney U statistic: 18247.0
P-value: 0.035084095693231204
Fail to reject null hypothesis: No significant difference between groups


Drawing from the results above, we could summarize Promotion_2 are significantly different with other promotion types, meanwhile Promotion_1 and Promotion_3 have no significant difference

## **Conclusion**

Taking the sales distribution into account where Promotion_2 sales are the lowest even with more market distribution employed. Promotion_3 has slightly higher total sales from Promotion_1 in four weeks  considered its market distribution also wider than Promotion_1

As a result, Promotion_1 or Promotion_3 with no significance difference can be considered as the best promotion to get implemented

We could look out to other details from both promotions such as cost production for Promotion_1 and Promotion_3 before choosing one of the promotions